bluesky/run_engine.py

```diff
@@ -1308,46 +1557,44 @@ def _run(self):
             # Emit RunStop if necessary.
             if self._run_is_open:
                 try:
-                    yield from self._close_run(Msg('close_run'))
+                    await self._close_run(Msg('close_run'))
                 except Exception:
                     self.log.error(
                         "Failed to close run %r.", self._run_start_uid)
 
             for p in self._plan_stack:
                 try:
                     p.close()
-                except RuntimeError as e:
+                except RuntimeError:
                     print('The plan {!r} tried to yield a value on close.  '
                           'Please fix your plan.'.format(p))
 
-            self.loop.stop()
             self._state = 'idle'
-        # if the task was cancelled
-        if pending_cancel_exception is not None:
-            raise pending_cancel_exception
-        self.log.info("Cleaned up from plan %r", self._plan)
 
-    @asyncio.coroutine
-    def _wait_for(self, msg):
-        """Instruct the RunEngine to wait until msg.obj has completed. Better
-        yet, see the docstring for ``asyncio.wait`` for what msg.obj should
-        be...
+        self.log.info("Cleaned up from plan %r", self._plan)
+        if isinstance(stashed_exception, asyncio.CancelledError):
+            raise stashed_exception
 
-        TODO: Get someone who knows how this works to check this note, since
-        it is almost assuredly total bs
+    async def _wait_for(self, msg):
+        """Instruct the RunEngine to wait for futures
 
         Expected message object is:
 
-            Msg('wait_for', None, futures, **kwargs)
+            Msg('wait_for', None, awaitable_factories, **kwargs)
+
+        The keyword arguments will be passed through to `asyncio.wait`.
+
+        The callables in awaitable_factories must have the signature ::
+
+           def fut_fac() -> awaitable:
+               'This must work multiple times'
 
-        Where ``obj`` and **kwargs are the position and keyword-only arguments
-        for ``asyncio.await``
         """
         futs, = msg.args
-        yield from asyncio.wait(futs, loop=self.loop, **msg.kwargs)
+        futs = [f() for f in futs]
+        await asyncio.wait(futs, loop=self.loop, **msg.kwargs)
```

Comment:
https://docs.python.org/3/library/asyncio-task.html#asyncio.wait
`The loop argument is deprecated and scheduled for removal in Python 3.10`

Reply:
Same comment as the other one, want to deal with this in a follow up PR as we need to sort out what the minimum version of python that correctly works without `loop` is.